src/Raven.Server/Rachis/Follower.cs

```diff
@@ -411,7 +411,74 @@ private void NegotiateWithLeader(TransactionOperationContext context, LogLengthN
             // the reason we send this is to simplify the # of states in the protocol
 
             var snapshot = _connection.ReadInstallSnapshot(context);
-            _debugRecorder.Record("Snapshot received");
+            _debugRecorder.Record("Start receiving snapshot");
+
+            // reading the snapshot from network and committing it to the disk might take a long time. 
+            using (var cts = new CancellationTokenSource())
+            {
+                var readAndCommitSnapshotTask = Task.Run(() => ReadAndCommitSnapshot(context, snapshot, cts.Token), cts.Token);
+                try
+                {
+                    WaitForTaskCompletion(readAndCommitSnapshotTask, "ReadAndCommitSnapshot");
+                }
+                catch
+                {
+                    cts.Cancel();
```

Comment:
In here, do we want to wait until `readAndCommitSnapshotTask` is done?
Because otherwise we might dispose the `cts` before it is noticed by the task

Reply:
done